### ISSUE TITLE
[DEVELOP] Roadmap for Integrating Pandoc & Deprecating 32bit

### ISSUE THREAD

**Issue by nathanlesage** [Zettlr/Zettlr#1552]
Dear all,

as promised, here's a short roadmap for how we'll proceed with migrating Zettlr from having to install Pandoc manually to Zettlr shipping with the binary.

## Current Status

Right now, everything is set up so that the CI toolchain will already bundle the pre-built Pandoc binaries with the appropriate installers (download them from the GitHub repository and then bundle them with the app). I've tested the packaging process out: everything works perfectly.

## Roadmap

- [x] **Release 1.8.4 (December 2020)**: Begin bundling the Pandoc binary with the 64bit builds (Windows, macOS, and Linux alike). If you have a 32bit build that does not contain Pandoc bundled, you will see a message indicating that. This means that you must upgrade your operating system, if you wish to continue using Zettlr. If you have a 64bit system and this message appears: that's a bug, please report! This does not apply to third-party builds, such as the Arch Linux release, which is being provided independent from me.
- [x] **Release 1.8.5 (January/February 2021)**: The app will begin using the bundled Pandoc version, if applicable. I will probably also shoot a message whether or not Zettlr has successfully used the internal Pandoc binary instead of your system. If that is the case, and you don't want Pandoc installed separately, you can at this point uninstall your own Pandoc binary, as Zettlr can successfully export without external tools. Please report if uninstalling your system wide Pandoc installation breaks Zettlr's exporter.
- [x] **Release 1.9.0 (February–April 2021)**: 32bit builds will be discontinued. Additionally, I plan to fully rebuild the exporter so that it's much more stable and can deal with Pandoc better. **This will also deprecate the Pandoc command, as it is very error prone and instead switch to a customizable UI in the preferences where you can turn the cogs of many supported options.**

> Please note that all times are tentative, and may be changed without further notice. Priority is that I can manage all necessary changes. To cite from the license (section 15., Disclaimer of Warranty):
> 
> THERE IS NO WARRANTY FOR THE PROGRAM, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES PROVIDE THE PROGRAM “AS IS” WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE PROGRAM IS WITH YOU. SHOULD THE PROGRAM PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR OR CORRECTION.

## Nice to know

- Pandoc's binary amounts to approximately 90 MB in size. This includes all templates, defaults, and whatnotelse. This means Zettlr's overall size will increase by this. Please keep in mind that Microsoft Word is over 2 GB in size, in case you are not satisfied with this.
- We discontinue 32bit builds because I neither have the time nor the motivation to set up a full Haskell build toolchain on our CI. Pandoc comes only for 64 bit systems. So if you are exporting using Zettlr, there's an almost 100 percent-chance your system will be 64 bit either way and you won't be affected by this.
- Do not attempt any PR at re-integrating 32bit builds by building such a Haskell toolchain. 32bit builds will never come back. The reasons for this are manifold, but the main is that 32bit is simply too old, and nobody should use this anymore. Not even companies. Any PR that attempts this will be closed without discussion, so if you want to contribute, please take a look at the open issue list; I am very happy with any help!
- Bundling Pandoc unfortunately does not (yet) free you from the necessity to manually install LaTeX, as PDF rendering will always require another application. However, I'm already looking around the net to see for viable solutions for all those of you who are happy with less powerful PDF engines.

**Comment by nathanlesage** [comment]
I'm gonna use **this comment (very likely to get edited as we go)** to document some other things we need to take care of.

**Feedback is very welcome, and if one or two of you want to step in with helping, stay tuned!**

## Pandoc variable resolution

Copied from @tarleb:

Direction of variable resolution:

1. metadata from defaults file (via --defaults)
2. metadata from dedicated metadata file (via --metadata-file)
3. metadata from command line (via --metadata or -M; repeating the same key creates an array)
4. metadata from the document (e.g., YAML blocks; keys in later blocks take precedence)

## Changes to the Pandoc command

The command will be removed in 1.9.0.

Instead, anything that can be steered using the GUI will be written into temporary defaults files which are passed to Pandoc instead, reducing the need for command line arguments as much as possible.

## Changes to the Preferences

The exporting tab should be improved so that you can steer the generation of documents much better without any command line shit going on. This effectively means to add some fancy buttons to control `defaults.yml`-variables.

The PDF-Preferences will completely just be written to the defaults files. This will also mitigate issues reported such as "please use the template referenced in the YAML frontmatter instead of the project template" and hence integrate Pandoc much more, making use of the neat stuff it offers.

**Comment by tarleb** [comment]
About invoking pandoc: Would you consider accepting a patch that adds an toggle such that users can take full control of the way exports are done? Should come with a big, fat warning how this is not advised and no support is given. The freedom would be *great* for advanced workflows (think Docker or Makefiles), but I wouldn't want to impose that on people potentially unfamiliar with the command line.

**Comment by nathanlesage** [comment]
> a big, fat warning

Like "by clicking this you accept to surrender your first-born to Jeff Bezos if you open one single issue because you mis-clicked"?

**Comment by the-solipsist** [comment]
From #1401:
> 
> * We definitely need to overwork the default templates Zettlr ships

Is there a substantial benefit from shipping default templates that are different from Pandoc's?

**Comment by nathanlesage** [comment]
Leaner. That was at least the benefit for the TeX templates. For HTML, it enables custom styling. But per se necessary not.

**Comment by tarleb** [comment]
> "by clicking this you accept to surrender your first-born to Jeff Bezos if you open one single issue because you mis-clicked"

We might need something stronger, as many firstborns are already claimed by various entities. But yeah, somewhere along these lines.

**Comment by Kangie** [comment]
> Leaner. That was at least the benefit for the TeX templates. For HTML, it enables custom styling. But per se necessary not.

I'd say that they're XeLaTeX specific, and actually easier for a layman to parse and understand than the default pandoc latex template. It could probably use better in-template documentation if we expect users to extend / customise it for their purposes.

It's **much** easier to customise the Zettlr template than the default pandoc template, as we **know** that we're using a KOMA-script document type and XeLaTeX, so we can avoid a whole bunch of nested $if$ statements (etc). 

I still think that It's a good idea to ship a leaner template with Zettlr and am happy to put some work into it. I've been ears-deep in latex for the last 6 months anyway.

**Comment by nathanlesage** [comment]
@Kangie Please, go forward. From now on you know what Pandoc version to expect, so please, if you could make that one (or two? Hehehe, we could ship multiples!) great default template, that'd be awesome ❤️

**Comment by Kangie** [comment]
Conveniently I just finished work on a command line / node port of the zettlr export-project-dir command with some extensions (front matter schema validation and revision history generated from git). 

I need to plumb that into some of my other projects, and develop templates from there, so I'll get my work back into Zettlr at some point.

I'll do up a couple of examples with catching degrees of complexity - one of my recent tricks has been defining fonts on a per unicode block basis to avoid tofu. Complex, but shows off the power of Unicode in XeLaTeX.

I'll also ensure that the title page code is in the template and enabled by a variable, which should reduce the complexity of the export command.

**Comment by Kangie** [comment]
See #1567 

Happy to implement any feedback. :)

**Comment by nathanlesage** [comment]
UPDATE: Release 1.8.5

We've progressed towards full integration of Pandoc. Pandoc will be bundled with 64 bit builds AND Apple Silicon builds.

## Question to Windows ARM users

Does Zettlr work normally exporting Zettlr using Pandoc 64 bit (Not! ARM 64)? If yes, please respond so that I can add Pandoc to the Windows ARM builds and continue supporting them!

***

Next steps:

1. Possibly add another Patch fixing the typing lag
2. Completely remove the Pandoc command mess
3. Instead, implement a general writer for `default.yaml` files
4. Adapt the settings towards supporting these as well
5. Possibly add a few more options
6. Test & Release 1.9 (or 2.0, to indicate that 32bit is finally history?)

**Comment by the-solipsist** [comment]
Hi @Kangie.  Thanks for putting in the work on improving the default LaTeX template for Zettlr.  My one suggestion: Keeping the changes from the Pandoc template to the minimum required, since the Pandoc default LaTeX template keeps getting updated and improved all the time; and the more we drift from Pandoc, the more of a pain I would imagine it would be to integrate those changes back into the Zettlr default LaTeX template.

**Comment by Kangie** [comment]
> My one suggestion: Keeping the changes from the Pandoc template to the minimum required, since the Pandoc default LaTeX template keeps getting updated and improved all the time; and the more we drift from Pandoc, the more of a pain I would imagine it would be to integrate those changes back into the Zettlr default LaTeX template.

I agree, to a point. 

The default pandoc template isn't particularly easy to parse for a layman, and it tries to cover the cases for  XeLaTeX, LaTeX, and LuaLaTeX to output nearly identical content independent of the TeX engine in use. It also tries to handle cases for using default document classes as well as the KOMA script ones.

Customising the template to this degree makes it (IMO) easier to understand, particularly for laymen. As we're bundling a pandoc version with the software, we don't need to worry too much about upstream breaking our template - we'll always have a chance to test when bumping the bundled pandoc version! We also get a simpler, easier to fix template because we don't have to parse all of those if statements ourselves!

I actually need to write some tests for my documentation build stuff (that this template is also used within) so that a CI build system can detect when document production breaks / changes due to pandoc updates, so I'll report back once that's done.

**Comment by the-solipsist** [comment]
Yep. Makes sense.  And once again, thanks for this.  (While I don't myself use Zettlr's export functionality, I know it would be very important for the vast majority of the users.)

> we don't need to worry too much about upstream breaking our template

I was thinking more about improvements from upstream having to be merged, rather than breakages caused by upstream.

**Comment by nathanlesage** [comment]
I agree with you both. I think, after all, those users who don't have any idea about LaTeX won't care how the template looks and those who do will be using their custom ones either way.

So I would suggest, we should orient ourselves along the following constraints:

- Require as few packages as necessary to minimize the packages required to install for newbie users
- Remove all the if else stuff because we're defaulting to XeLaTeX and KOMA either way as we're targeting these specific use cases
- Make it so that the following types of documents can be easily generated using Zettlr and Pandoc:
    - Papers
    - Simple documents
    - book-like prints

Ideally all of this using 100% Pandoc compatible things do that we can control it easily. I will begin writing a defaults YAML writer soon and can implement any switches necessary for this!

**Comment by nathanlesage** [comment]
Ah, and @Kangie you mentioned exporting tests? If that is something we can run on a CI, we could implement a workflow for this here!

**Comment by Kangie** [comment]
>     * Require as few packages as necessary to minimize the packages required to install for newbie users

The template won't require anything not installed by the TeX Live packages (for Linux)

On Windows, MikTeX will actually download any missing packages from CTAN on first use with no interaction, so no issues using any package there. 

I'm not sure how MacTeX handles things, but I assume that it's pretty similar to TeX Live in terms of packages.

>    * Remove all the if else stuff because we're defaulting to XeLaTeX and KOMA either way as we're targeting these specific use cases

Already done.

>     * Make it so that the following types of documents can be easily generated using Zettlr and Pandoc:
>       
>       * Papers
>       * Simple documents
>       * book-like prints

I've got a Beamer template nearly ready, and most of the other things can be accomplished by adding a few more variables to the one that's already in #1567. It should, in theory, be possible to switch between most of those just by adjusting the document class. If that's the case we may as well add letter to the list and cover all of the KOMA classes:

- Article
- Book
- Report
- Letter
- Beamer

`scrartcl` should be the default  class - it best translates the concepts that users are familiar with in a WYSIWYG editor (e.g. Word, LibreOffice Writer) onto the resulting PDF.

It may be worth implementing some sort of frontmatter validation - something like [this](https://github.com/Lackadaisical-Documentation-Generator/lackadaisical-markdown-binder/blob/master/util/content/parse-frontmatter.js), perhaps? As a bonus it's capable of extracting the frontmatter to a pandoc yaml metadata-file if that's desirable.

I can put together a schema for the non-boolean values when all of the templates are done.

I know that you've been against the idea, but are you interested in a word / powerpoint reference doc? I recently had to learn how to create them, particularly custom table styles!

Do you have any thoughts on colours? I'm currently just using the Zettlr Logo's green as a colour on the title page, but may also enable  optional coloured headings. What would you like the default to be? Should this be configurable (aside from toggling on and off), or would we expect the user to provide a modified template? At some point we'll need to draw the line.

**Comment by Kangie** [comment]
> Ah, and @Kangie you mentioned exporting tests? If that is something we can run on a CI, we could implement a workflow for this here!

That's entirely the purpose. It's reasonably high on my priority list, so I'm hopeful to get started on it sometime this month.

**Comment by nathanlesage** [comment]
> The template won't require anything not installed by the TeX Live packages (for Linux)
>
> On Windows, MikTeX will actually download any missing packages from CTAN on first use with no interaction, so no issues using any package there.
>
> I'm not sure how MacTeX handles things, but I assume that it's pretty similar to TeX Live in terms of packages.

I actually had an interaction on a Windows computer a few years ago ("Package xyz is required, do you want to download it?"), and I know that the default template had some problems on MacTex as well because there were a few packages that were required …

But if that's gone now, all the better. I will have the chance to test this next week or so myself, because I'm waiting for a brand new MacBook, which doesn't have TeX installed yet.

> Already done.

🤝 

> It should, in theory, be possible to switch between most of those just by adjusting the document class. If that's the case we may as well add letter to the list and cover all of the KOMA classes

Sure, what I mentioned was only what MUST be supported, everything above that all the better :)

> It may be worth implementing some sort of frontmatter validation - something like this, perhaps? As a bonus it's capable of extracting the frontmatter to a pandoc yaml metadata-file if that's desirable.

You mean something like [this](https://github.com/Zettlr/Zettlr/blob/develop/source/common/util/extract-yaml-frontmatter.js) in conjunction with [that](https://github.com/Zettlr/Zettlr/blob/develop/source/main/modules/fsal/fsal-file.ts)? I must agree that I just now realized that the frontmatter validation should be moved to its own utility function now that I see that, but in general, I've been adding more and more validation, because stupid sh** like "Ah, yeah, it's an empty property, so we will define the property on the resulting object and set it to an empty string instead of leaving it undefined", see for that [this comment](https://github.com/Zettlr/Zettlr/blob/235828c0ce1904173a89245ed3fb946358cd84d5/source/main/modules/fsal/fsal-file.ts#L148-L160).

> I can put together a schema for the non-boolean values when all of the templates are done.

This would indeed quickstart the implementation of a bunch of Pandoc variables, I can extend Zettlr's internal validator for this (it's been neglected far too long either way)

> I know that you've been against the idea, but are you interested in a word / powerpoint reference doc? I recently had to learn how to create them, particularly custom table styles!

Whaaaaaaat, I have not been against this idea :o I've done this in the past (Zettlr had a few versions where it shipped with custom reference docs, actually :D), and there is this issue open which demands a GUI option to set reference docs, which I plan on implementing alongside trashing this ugly Pandoc command thingy! I would then say, we could ship one default reference doc, and I'll implement a way for users to simply add a reference doc from somewhere else, in case they want another thing.

> Do you have any thoughts on colours? I'm currently just using the Zettlr Logo's green as a colour on the title page, but may also enable optional coloured headings. What would you like the default to be? Should this be configurable (aside from toggling on and off), or would we expect the user to provide a modified template? At some point we'll need to draw the line.

In general, my current goal is the following:

- For everything, enable Zettlr to immediately do this by itself without further configuration, so basically the macOS way: Ship with sane defaults.
- As soon as the user wants to modify this, the following two rules apply:
- Either it can be set using a Pandoc variable. If that is the case, it'll be implemented by me in the preferences, so that the options can be written to the defaults-file
- If this cannot be set using a Pandoc variable (e.g. the guts of a template), we can provide a way for users to use different templates.

So, TL;DR: No user-modification of the built-in templates that goes below Pandoc-supported variables. For everything else, I expect users to provide more templates. Maybe we could even create some sort of "market"place for users to upload and share these things, but that's another story.

> Edit: I'm pretty fond of this, actually.

This palette feels pretty … traditionalist? :D 

> That's entirely the purpose. It's reasonably high on my priority list, so I'm hopeful to get started on it sometime this month.

Looking forward to it ❤️

**Comment by Kangie** [comment]
> I actually had an interaction on a Windows computer a few years ago ("Package xyz is required, do you want to download it?")

I may have overstated that - Windows users won't need to fetch the packages manually, but they may still be prompted. It's been more than 6 months since I seriously used MiKTeX!

> But if that's gone now, all the better. I will have the chance to test this next week or so myself, because I'm waiting for a brand new MacBook, which doesn't have TeX installed yet.

Great. We'll see how it goes. Some options (particularly the default font configuration) work really well in my docker-based documentation production setup, but may be a little bit of overkill for most users. I'll provide some instructions for testing before we tweak the templates and settle on a final design. 

My aim is to have a sane set of defaults that look professional but that can be easily amended through variables (where possible), and be easy to understand, extend and maintain!

> This palette feels pretty … traditionalist? :D

I am 100% open to suggestions for default colours (and honestly other design aspects!). I just hit generate until the colours looked "good" to me, particularly in the contexts of 'adding a splash of colour' to documents / presentations.

It's also for that reason that our vector title backgrounds / slide backgrounds all features triangles: to misquote Star Trek: I'm a tech writer, not a designer!

I'll get a pptx/potx (ew) reference document together over the next week or so. It's nowhere near as good at layouts as the Beamer templates, but we can at least get consistent coloured table rows and designs between the two. The PowerPoint export handles presenter notes better than beamer at this point, but the table style has to be hand-coded; providing one with sane defaults will make it much easier as users get to find-and-replace HTML/RGB colour values instead of learning to write valid ooxml (particularly when the Microsoft-provided validator doesn't work!).

I'm working on the beamer template at the moment. Aside from some issues with the title slide (I think that there's a 'frame' taking up the bottom 1/3 of the slide beneath my 'bottom-aligned' title content!) it's looking pretty good. I also have some improvements to the template that's already in my PR, so keep an eye out. 

I'll provide some sample content that you can test with (to see how wrong I am about packages) and strip the frontmatter options from to try with arbitrary content, along with a list of caveats.

**Comment by nathanlesage** [comment]
> to misquote Star Trek: I'm a tech writer, not a designer!

:D I feel you there. I can ask around if I find some designer that might give us a PDF including better styling which we can then implement!

My MacBook has just been shipped, so just a few days now until I can test the TeX stuff!

**Comment by tarleb** [comment]
Maybe the fine folks at https://opensourcedesign.net can help with the design.

**Comment by nathanlesage** [comment]
UPDATE to the migration:

***

We've had some reports that the bundled Pandoc in 1.8.5 doesn't work on Windows (and on some Unix systems as well if the username includes a space). The reason is as silly as bad: We are using node's `child_process.exec` to execute Pandoc including the Pandoc command. To use the bundled Pandoc, we need to pass the full, absolute Path to the executable. The reason for this is that I couldn't find *any* information on how Node is using `process.env.PATH` to resolve binaries, so I cannot be sure if it suffices to stick the resources path into the environment PATH at the first position to reliably call the bundled Pandoc, and not some potential system version which might or might not be outdated. This will hold true for all time, because we will always encounter some systems where some Pandoc version might be installed, so we need to pass the absolute path.

However, `exec` just runs a shell, passing the first parameter as if we had written it manually. If the path therefore includes a space, this will break the command (because `C:\Program` is not actually a binary).

This means that we will have to wait for the 1.9 betas to fully test this out, because I have to switch to `spawn`, which takes a path argument that can contain spaces. On Windows, it will never work because, for instance, `Program Files` contains a space. And that's where Zettlr is installed. I cannot implement `spawn` before trashing the Pandoc command, because it is unsafe to just split by space and might lead to unintended consequences. I'd rather play it safe here.

**Comment by Kangie** [comment]
Good luck conscripting some 'volunteer' designers.

I'm still working on a module that returns pandoc defaults file objects. I feel like I have most of the code in place, and I'm just finishing up some tests... so that means it'll be another few weeks. :p

I'm hopeful that it'll be ready to go by the weekend. Writing a basic JSONSchema for the defaults file was an endeavour.

**Comment by nathanlesage** [comment]
> I'm hopeful that it'll be ready to go by the weekend. Writing a basic JSONSchema for the defaults file was an endeavour.

:)

> Edit: Can you just quote the path?

I don't think so; apparently quoting strings on a Windows shell involves some very ugly control characters, but I'm not sure <.<

P.S.: Regarding getting get-pandoc.sh to run on Windows might involve using the Bash, I adapted the script several times so that it should work fine on anything that runs bash syntax (I've had some problems on the macOS CI because I accidentally used Shell syntax first … this is some first-grade inconsistency, I tell ya)

**Comment by Kangie** [comment]
I guess that we can just recommend cygwin for windows development. I had it installed and didn't even consider it

**Comment by nathanlesage** [comment]
cygwin …?

**Comment by nathanlesage** [comment]
## Update to the Migration

I just fiddled around with the PATH environment variable and now it's working like this (also live changes for 1.8.6):

1. If Pandoc has been bundled, the environment variable `PANDOC_PATH` will be set and not `undefined`
2. The `PANDOC_PATH` will be *unshifted* in front of the `process.env.PATH` so that this directory will be searched **first**. But only (!) if you selected the option to use the bundled Pandoc.
3. If you switch the option (whether or not the bundled Pandoc will be used), you will now have to reboot the application, because `PANDOC_PATH` is being patched on application boot.

Please, all of you: Test out whether or not this behaviour works on Windows and Linux as well, but I've researched for one week now and this is the solution based upon all the amassed knowledge!

**Comment by nathanlesage** [comment]
## Update to the Migration

1. I just trashed the Pandoc command for good
2. The exporter is now written in TypeScript
3. It is using a defaults file
4. Switched the Pandoc calling logic to `spawn`
5. Migrated most of the settings to the defaults writer
6. Some cleanup with regard to the revealJS builder and other files
7. A lot of files have been inappropriately touched today.

**Comment by nathanlesage** [comment]
Please also refer to the now-main issue #1690.

**Comment by nathanlesage** [comment]
Since we have already deprecated 32bit and Pandoc sits nice and tightly in the application package, I'm gonna close this issue finally :)
